File: infra-market-server/src/main/kotlin/io/infra/market/service/TokenService.kt
package io.infra.market.service

import io.infra.market.util.JwtUtil
import io.infra.structure.redis.core.JedisTemplate
import org.springframework.stereotype.Service

@Service
class TokenService(
    private val jedisTemplate: JedisTemplate
) {
    
    private val tokenPrefix = "token:"
    private val tokenExpireTime: Long = 3 * 24 * 60 * 60 // 3天（秒）
    
    /**
     * 保存token到Redis
     */
    fun saveToken(uid: Long, token: String) {
        val key = "$tokenPrefix$uid"
        jedisTemplate.setex(key, tokenExpireTime, token)
    }
    
    /**
     * 从Redis获取token
     */
    fun getToken(uid: Long): String? {
        val key = "$tokenPrefix$uid"
        return jedisTemplate.get(key)
    }
    
    /**
     * 验证token是否有效
     */
    fun validateToken(token: String): Boolean {
        val uid = JwtUtil.getUidFromToken(token) ?: return false
        
        // 验证JWT token本身
        if (!JwtUtil.validateToken(token)) {
            return false
        }
        
        // 验证Redis中的token
        val storedToken = getToken(uid)
        return storedToken == token
    }
    
    /**
     * 删除token
     */
    fun deleteToken(uid: Long) {
        val key = "$tokenPrefix$uid"
        jedisTemplate.del(key)
    }
    
    /**
     * 刷新token
     */
    fun refreshToken(uid: Long, username: String): String {
        val newToken = JwtUtil.generateToken(uid, username)
        saveToken(uid, newToken)
        return newToken
    }
}

File: infra-market-server-go/internal/service/token_service.go
package service

import (
	"context"
	"fmt"
	"time"

	"github.com/bucketheadv/infra-market/internal/util"
	"github.com/go-redis/redis/v8"
)

type TokenService struct {
	redisClient *redis.Client
}

func NewTokenService(redisClient *redis.Client) *TokenService {
	return &TokenService{redisClient: redisClient}
}

const (
	tokenPrefix     = "token:"
	tokenExpireTime = 3 * 24 * 60 * 60 // 3天（秒）
)

// SaveToken 保存token到Redis
func (s *TokenService) SaveToken(uid uint64, token string) {
	key := fmt.Sprintf("%s%d", tokenPrefix, uid)
	s.redisClient.Set(context.Background(), key, token, time.Duration(tokenExpireTime)*time.Second)
}

// GetToken 从Redis获取token
func (s *TokenService) GetToken(uid uint64) (string, error) {
	key := fmt.Sprintf("%s%d", tokenPrefix, uid)
	return s.redisClient.Get(context.Background(), key).Result()
}

// ValidateToken 验证token是否有效
func (s *TokenService) ValidateToken(token string) bool {
	uid, err := util.GetUIDFromToken(token)
	if err != nil {
		return false
	}

	// 验证JWT token本身
	if !util.ValidateToken(token) {
		return false
	}

	// 验证Redis中的token
	storedToken, err := s.GetToken(uid)
	if err != nil {
		return false
	}

	return storedToken == token
}

// DeleteToken 删除token
func (s *TokenService) DeleteToken(uid uint64) {
	key := fmt.Sprintf("%s%d", tokenPrefix, uid)
	s.redisClient.Del(context.Background(), key)
}

// RefreshToken 刷新token
func (s *TokenService) RefreshToken(uid uint64, username string) (string, error) {
	token, err := util.GenerateToken(uid, username)
	if err != nil {
		return "", err
	}
	s.SaveToken(uid, token)
	return token, nil
}
